AI Workspaces

enisn opened the pull request:
### Description

Introduce `Volo.Abp.AI` with `Microsoft.Extensions.AI` and `SemanticKernel` integration by using Workspaces


### Checklist

- [x] I fully tested it as developer / designer and created unit / integration tests
- [x] I documented it (or no need to document or I will create a separate documentation issue)

hikalkan committed:
Created the Volo.Abp.AI package

hikalkan committed:
Add Microsoft.Extensions.AI package reference

hikalkan committed:
Initial AbpAIOptions definition

hikalkan committed:
Implement builders

hikalkan committed:
Refactor the client building code

hikalkan committed:
Rename AbpAIOptions to AbpAIChatClientOptions

hikalkan committed:
Update AbpAIModule.cs

hikalkan committed:
Introduce typed chat client services

hikalkan committed:
Refactored AI services

hikalkan committed:
Register IChatClient for the default configuration

hikalkan committed:
Update AbpAIModule.cs

hikalkan committed:
Update ChatClientConfigurationCollection.cs

enisn committed:
Merge branch 'dev' into ai

enisn committed:
Refactor ChatClientNameAttribute to reduce reflection calls

enisn committed:
Update Microsoft.Extensions.AI package to version 9.7.1, add Microsoft.Extensions.AI.Abstractions package, and introduce new abstractions for AI functionality. Include FodyWeavers configuration and schema files for weaving support.

enisn committed:
Add AI chat client delegates and extension methods

Introduced new delegates for ChatClient with system message, temperature, and tools in the Volo.Abp.AI.Abstractions project. Added corresponding extension methods in Volo.Abp.AI. Updated the solution file to include the new project and configuration platforms.

enisn committed:
Add ChatClientWithSystemMessage abstraction and extensions

Introduced the Volo.Abp.AI.Abstractions project with a new ChatClientWithSystemMessage delegate and related extension methods. Updated the solution and AbpAIModule to include and reference the new abstractions, supporting enhanced AI chat client extensibility.

enisn committed:
Merge branch 'ai' of https://github.com/abpframework/abp into ai

enisn committed:
Add SemanticKernel support

enisn committed:
Expose InnerKernel with a property

enisn committed:
Create README.md

enisn committed:
Refactor AI options to use workspace-based configuration

Replaces separate ChatClient and Kernel configuration dictionaries with a unified WorkspaceConfigurationDictionary. Removes related collection classes and extensions, updates service registration logic, and renames IKernel to IKernelAccessor. This change simplifies configuration and improves extensibility for managing multiple AI workspaces.

enisn committed:
Refactor workspace config to use workspace name property

Updated WorkspaceConfiguration to include a Name property and refactored service registration to use workspaceConfig.Name instead of nested ChatClient/Kernel names. Simplified ChatClientConfiguration and KernelConfiguration by removing the Name property and constructors, and added fluent configuration methods to WorkspaceConfiguration for improved usability.

enisn committed:
Refactor WorkspaceConfiguration property initialization

Changed ChatClient and Kernel properties in WorkspaceConfiguration to be non-nullable and initialized by default. Removed null checks and redundant instantiation logic in configuration methods for improved code clarity and safety.

enisn committed:
Merge branch 'dev' into ai

enisn committed:
Merge branch 'dev' into ai

enisn committed:
Implement Workspace Configuration Classes and Typed Kernel Accessor

Added new classes: WorkspaceConfiguration, WorkspaceConfigurationDictionary, and TypedKernelAccessor. These classes facilitate workspace-specific configurations and provide a typed access mechanism for the kernel service. Removed the previous implementation of WorkspaceConfiguration and its dictionary to streamline the configuration process.

enisn committed:
Merge branch 'dev' into ai

enisn committed:
Merge branch 'dev' into ai

enisn committed:
Separate PreConfigure & Configure logic

enisn committed:
Allow nullable configureAction parameters.

enisn committed:
Allow null configuration in AbpAIOptions

enisn committed:
Update AbpAIModule.cs

enisn committed:
Merge branch 'dev' into ai

enisn committed:
Remove ChatClientWithTools due it's already possible by configuring default ME I

enisn committed:
Initial AI documentation

hikalkan commented on framework/src/Volo.Abp.AI.Abstractions/Volo/Abp/AI/Delegates/ChatClientWithSystemMessage.cs:
```diff
@@ -0,0 +1,52 @@
+
+namespace Volo.Abp.AI.Delegates;
+
+public class ChatClientWithSystemMessage : DelegatingChatClient
```
Best to remove this class to keep this package minimal.

hikalkan commented on docs/en/framework/infrastructure/artificial-intelligence.md:
```diff
@@ -0,0 +1,311 @@
+# Artificial Intelligence
```
Completely review this document, maybe rewrite some parts

hikalkan commented on framework/src/Volo.Abp.AI.Abstractions/Volo/Abp/AI/ChatClientNameAttribute.cs:
```diff
@@ -0,0 +1,38 @@
+
+namespace Volo.Abp.AI;
+
+[AttributeUsage(AttributeTargets.Class | AttributeTargets.Interface | AttributeTargets.Struct)]
```
Only class is enough.

hikalkan committed:
Remove unnecessary package dependencies

hikalkan commented on framework/src/Volo.Abp.AI.Abstractions/Volo.Abp.AI.Abstractions.csproj:
```diff
@@ -0,0 +1,26 @@
+        <ProjectReference Include="..\Volo.Abp.Core\Volo.Abp.Core.csproj" />
+    </ItemGroup>
+
+    <ItemGroup>
```
I removed unnecessary package dependencies: https://github.com/abpframework/abp/commit/91bbf064ee7d163c1a8a7143460370f07c1d4f3d

hikalkan commented on framework/src/Volo.Abp.AI/Volo/Abp/AI/Extensions/ChatClientWithSystemMessageExtensions.cs:
```diff
@@ -0,0 +1,12 @@
+
+public static class ChatClientWithSystemMessageExtensions
+{
+    public static ChatClientBuilder UseSystemMessage(this ChatClientBuilder builder, string systemMessage)
```
Remove this

enisn committed:
Refactor system message handling and update docs

Removed ChatClientWithSystemMessage and its extension, consolidating system message logic into user-extensible patterns. Renamed ChatClientNameAttribute to WorkspaceNameAttribute and restricted its usage to classes. Added ConfigureDefault to WorkspaceConfigurationDictionary for easier default workspace setup. Updated documentation to reflect these changes and clarify configuration patterns.
